### PR TITLE
[Gmail] New Components

### LINKED ISSUE #4064: Actions and Triggers for Gmail
### App
`gmail`
https://developers.google.com/gmail/api

### Triggers
- New email (include an optional prop to specify a search criteria)
- New attachment
- New email with label

### Actions
- Add label to email (async options props for Labels and Message)
- Find email (prop for "Search string", which should use Gmail's search)
- Send email

Send email action should include the below props:
- To (type: `string[]`, optional)
- Cc (type: `string[]`, optional)
- Bcc (type: `string[]`, optional)
- From Name (optional)
- Reply To (optional)
- Subject (required)
- Body Type (plain or html; optional; default to plain)
- Body (required, type: any)
- Signature (optional, list via async options)
- Attachments (optional, string)

### PR DESCRIPTION
**Pull request by andrewjschuang**
New components for Gmail

**Sources:**
- new-email (all emails or filter by label and/or search query)
- new-attachment

**Actions:**
- add-label-to-email
- find-email
- send-email

**Note:**

Webhooks are not available for Gmail since the only way supported is through Google Cloud Pub/Sub which brings in a lot more steps for the user and needs to enable billing. More information [here.](https://developers.google.com/gmail/api/guides/push)

Resolves #4064.

<a href="https://gitpod.io/#https://github.com/PipedreamHQ/pipedream/pull/4100"></a>

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

**Review by feyzullah — APPROVED** [review]

**Comment by dannyroosevelt** [comment]
Nice, @andrewjschuang!

For the Send Email action, I think we are using the email in the `to` prop, for the `bcc` prop. Is that possible?

**Comment by dannyroosevelt** [comment]
And @andrewjschuang would the `attachments` prop make more sense as an object, where the `key` is the filename and the `value` is the url, instead of having users concat those as a string?

File: components/gmail/actions/send-email/send-email.mjs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    signature: {
+      propDefinition: [
+        gmail,
+        "signature",
+      ],
+    },
+    attachments: {
+      type: "string[]",
+      label: "Attachments",
+      description: "A list of attachments. Each attachment sould be a `filename,url` formatted string. The `filename` should contain the file extension (i.e. `.jpeg`, `.txt`) and the `url` is the download link for the file.",
```
Comment (dannyroosevelt) [inline components/gmail/actions/send-email/send-email.mjs:66]: Typo: `sould` --> `should`

File: components/gmail/actions/send-email/send-email.mjs [hunk 2]
```diff
@@ -0,0 +1,110 @@
+/* eslint-disable pipedream/props-description */
+import gmail from "../../gmail.app.mjs";
+import constants from "../../common/constants.mjs";
+
+export default {
+  key: "gmail-send-email",
+  name: "Send Email",
+  description: "Send an email",
```
Comment (dannyroosevelt) [inline components/gmail/actions/send-email/send-email.mjs:8]: "Send an email from your Google Workspace email account."

File: components/gmail/gmail.app.mjs [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        return sendAs.map(({ signature }) => ({
+          label: convert(signature),
+          value: signature,
+        }));
+      },
+    },
+    q: {
+      type: "string",
+      label: "Search Query",
+      description: "Search query used in Gmail. [More information here](https://support.google.com/mail/answer/7190).",
```
Comment (dannyroosevelt) [inline components/gmail/gmail.app.mjs:68]: Applies a search filter using Gmail's [standard search operators](https://support.google.com/mail/answer/7190).

**Comment by dannyroosevelt** [comment]
@andrewjschuang 
Screenshot by Dropbox Capture

**Comment by dannyroosevelt** [comment]
@andrewjschuang let's also add an additional option prop to the Send Email action:
- Send Email From (should list any other emails the user can send from, I think these are "delegate" emails)

**Comment by andrewjschuang** [comment]
@dannyroosevelt all your comments were addressed. Thanks!

**Comment by dannyroosevelt** [comment]
If this prop is unset, what does the API do? (I think we should probably default to **false**)
Screenshot by Dropbox Capture

**Comment by dannyroosevelt** [comment]
I just pushed some small tweaks to a couple prop descriptions

**Comment by dannyroosevelt** [comment]
Btw awesome job on these, @andrewjschuang!

**Comment by andrewjschuang** [comment]
> If this prop is unset, what does the API do? (I think we should probably default to **false**)

Yep, it defaults to `false`. Added this to the description.

**Comment by vunguyenhung** [comment]
Hello @andrewjschuang, I have tested this PR and some improvements are needed for `Find Email` and `Send Email`. All other actions & sources have passed!

Please check the test report below for more information
[Gmail_4100_2149.pdf](https://github.com/PipedreamHQ/pipedream/files/9445489/Gmail_4100_2149.pdf)

**Comment by vunguyenhung** [comment]
Hi everyone, all test cases are passed! Ready for release! 

Test report
[Gmail_4100_2314.pdf](https://github.com/PipedreamHQ/pipedream/files/9446177/Gmail_4100_2314.pdf)

**Comment by andrewjschuang** [comment]
@dannyroosevelt do you see anything else before merging?

**Review by lcaresia — APPROVED** [review]
LGTM!

**Pull request merged**